### PR TITLE
feat(robot-server): cal check mock api

### PR DESCRIPTION
**Pull request by amitlissack**
## overview

This is a mocked up calibration check http api implementation. 

These endpoints will be the first to utilize the jsonapi based v2 schema Ian created.  This PR represents the way will define our API going forward.  

## review requests

- Ignore that tests don't pass. 

- Think deeply not just about the implications for calibration and sessions, but for all future API. Do we like that all responses have this schema: 
```
{
  "meta": {},
  "links": {
     "nextstep": "/someurl/12345/go"
  },
  "data": {
    "id": "id of object",
    "type": "type represented in attributes",
    "attributes": {
      "name": "string",
      "quantity": 0,
      "price": 0,
      .......
    }
  }
}
```

- Best way to review is to run the robot server and view the docs:
  - run it `OT_API_FF_useFastApi=TRUE make dev`
  - go to http://localhost:31950/redoc or http://localhost:31950/docs#/
  - Look at the `Sessions` and `Calibration Check` sections.

## risk assessment

None.

### CODE REVIEW

File: robot-server/robot_server/service/routers/session.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+from robot_server.service.models import session
+from robot_server.service.models.json_api.response import ResponseModel, \
+    ResponseDataModel
+
+SessionResponseData = ResponseDataModel[session.Session]
+SessionResponse = ResponseModel[SessionResponseData]
+MultiSessionResponse = ResponseModel[typing.List[SessionResponseData]]
+
+router = APIRouter()
+
```
Comment (amitlissack) [inline robot-server/robot_server/service/routers/session.py:14]: There will come a time where there will be multiple kinds of sessions. Let's not use calibration or check in the create, delete calls.

File: robot-server/robot_server/service/routers/calibration_check.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+
+StatusDataModel = response.ResponseDataModel[models.CalibrationSessionStatus]
+StatusModel = response.ResponseModel[StatusDataModel]
+
+RequestDataModel = request.RequestDataModel[CheckCommand]
+RequestModel = request.RequestModel[RequestDataModel]
+
+
+@router.post("/session/{session_id}/check_command",
```
Comment (amitlissack) [inline robot-server/robot_server/service/routers/calibration_check.py:37]: There may be multiple sessions by type.

We could consider omitting the {**session_id**} in the path as it is also present in the body of the message. That would gave us the luxury of creating sessions on demand.

File: robot-server/robot_server/service/routers/calibration_check.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+             response_model=StatusModel)
+async def calibration_command(session_id: str,
```
Comment (amitlissack) [inline robot-server/robot_server/service/routers/calibration_check.py:39]: Do we want command endpoints to return a standard response for all commands regardless of the session type?

Perhaps the session specific stuff should be in the `metadata` parameter? For example, calibration check returns a large blob of data about the calibration check session state.

File: robot-server/robot_server/service/routers/calibration_check.py (hunk 2 above)
Comment (mcous) [inline robot-server/robot_server/service/routers/calibration_check.py:37]: Given the option between

```
POST /session/:session_id/command_type_1 { ...commandData }
POST /session/:session_id/command_type_2 { ...commandData }
POST /session/:session_id/command_type_3 { ...commandData }
```

...and...

```
POST /session/:session_id/commands/execute { commandType: command_type_1, ...commandData }
POST /session/:session_id/commands/execute { commandType: command_type_2, ...commandData }
POST /session/:session_id/commands/execute { commandType: command_type_3, ...commandData }
```

I would _really_ prefer option 2. 

- Eventually, we should be able to create a command without necessarily executing it or blocking the response until it completes
    - A command is a resource that we create, robot execution based on that resource is an asynchronous side-effect
- An endpoint that executes a given command and blocks the response is an important convenience feature

I'd been imagining something like:

```
POST /session/:session_id/commands { commandType: ... }
> { type: 'command', data: { id: 'abc', status: 'queued', ... } }

GET /session/:session_id/commands/:command_id
> { type: 'command', data: { id: 'abc', status: 'executing', ... } }

POST /session/:session_id/commands/execute { commandType: ... }
> { type: 'command', data: { id: 'def', status: 'done', ... } }
```

File: robot-server/robot_server/service/routers/session.py (hunk 1 above)
Comment (mcous) [reply robot-server/robot_server/service/routers/session.py:14]: Related to my other comments about commands: do different session types need their own routers? Our generic session routes could include generic command endpoints and each session type could define implementations for whatever commands (based on session type, session state, etc.)

**Review by mcous — COMMENTED** [review]
I'm liking the direction this is heading. Left my two cents

> - Think deeply not just about the implications for calibration and sessions, but for all future API. Do we like that all responses have this schema: 
> ```
> {
>   "meta": {},
>   "links": {
>      "nextstep": "/someurl/12345/go"
>   },
>   "data": {
>     "id": "id of object",
>     "type": "type represented in attributes",
>     "attributes": {
>       "name": "string",
>       "quantity": 0,
>       "price": 0,
>       .......
>     }
>   }
> }
> ```

I've felt a _little_ iffy on the `data` and `attributes` objects in the past; they feel a little bit like unnecessary nesting to me. I could be talked into flattening one or both of them into a response interface that includes `meta`, `links`, `type`, `id` (if applicable), and then whatever arbitrary model fields. We'd have to carefully consider collection responses if we go the flattening route, though (perhaps with a `type: 'collection'`).

File: robot-server/robot_server/service/routers/session.py (hunk 1 above)
Comment (amitlissack) [reply robot-server/robot_server/service/routers/session.py:14]: Yes. Indeed.

File: robot-server/robot_server/service/routers/calibration_check.py (hunk 2 above)
Comment (amitlissack) [reply robot-server/robot_server/service/routers/calibration_check.py:37]: Ok, so perhaps we omit the `session_id` in the path. We will make a decision based on the `id` in the payload.

We can make a decision on the backend if the command can use a default session.

FWIW, I don't think calibration check should be a default session type.

File: robot-server/robot_server/service/routers/calibration_check.py (hunk 3 above)
Comment (amitlissack) [reply robot-server/robot_server/service/routers/calibration_check.py:39]: The more I think about it the more I feel like we can put session specific state stuff in the `metadata` do you agree with that?

**Comment by amitlissack** [comment]
> I'm liking the direction this is heading. Left my two cents
> 
> > * Think deeply not just about the implications for calibration and sessions, but for all future API. Do we like that all responses have this schema:
> > 
> > ```
> > {
> >   "meta": {},
> >   "links": {
> >      "nextstep": "/someurl/12345/go"
> >   },
> >   "data": {
> >     "id": "id of object",
> >     "type": "type represented in attributes",
> >     "attributes": {
> >       "name": "string",
> >       "quantity": 0,
> >       "price": 0,
> >       .......
> >     }
> >   }
> > }
> > ```
> 
> I've felt a _little_ iffy on the `data` and `attributes` objects in the past; they feel a little bit like unnecessary nesting to me. I could be talked into flattening one or both of them into a response interface that includes `meta`, `links`, `type`, id`(if applicable) ,and then whatever arbitrary model fields. We'd have to carefully consider collection responses if we go the flattening route, though (perhaps with a`type: 'collection'`).

I can't see the `id` being at the root level if we support collections. Unless you're thinking we return an object or a collection (?)

File: robot-server/robot_server/service/routers/calibration_check.py (hunk 2 above)
Comment (mcous) [reply robot-server/robot_server/service/routers/calibration_check.py:37]: Sorry, I meant to say that the session_id could be optional, but I do think that if we're in a situation where we have an ID'd session, I'd rather have the ID in the path than in the body. Why is the session ID present in the body here?

Agreed, calibration check for sure shouldn't be default. In the past I had imagined the default as a session type that couldn't do anything with deck or pipette state, so if you need to do anything more complicated you need to explicitly create something

**Comment by mcous** [comment]
> I can't see the id being at the root level if we support collections. Unless you're thinking we return an object or a collection (?)

Yeah an `id` it wouldn't make sense in that context. If we wanted to flatten a little and be consistent with collections, I _guess_ I'm proposing something along the lines of:

```js
{
  // ...
  type: 'collection',
  collection: [
    {
      type: 'item'
      id: 'abc',
      // ...
    },
    // ...
  ]
}
```

File: robot-server/robot_server/service/routers/calibration_check.py (hunk 2 above)
Comment (amitlissack) [reply robot-server/robot_server/service/routers/calibration_check.py:37]: I am in total agreement.

File: robot-server/robot_server/service/routers/calibration_check.py (hunk 3 above)
Comment (amitlissack) [reply robot-server/robot_server/service/routers/calibration_check.py:39]: What I'd like to do is "genericize" the `meta` field. Like Ian did with `attributes`. It means that the server can define pydantic models for the `meta` field that are specific to session type.

File: robot-server/robot_server/service/routers/calibration_check.py (hunk 2 above)
Comment (amitlissack) [reply robot-server/robot_server/service/routers/calibration_check.py:37]: One challenge on the server side will be to define pydantic Models for commands.  

We can do something like defining the basic structure common to all commands with specific stuff being formless. 

```
class SessionCommand(BaseModel):
   command_type: CommandType
   id: str
   body: Dict[str, Any]
```

or something like this where the body is a Union of all possible command models.

```
class PipetteCommand(BaseModel):
    mount: Mount

class JogCommand(PipetteCommand):
   vector: List[float]


class SessionCommand(BaseModel):
   command_type: CommandType
   id: str
   body: Union[JogCommand, PipetteCommand]
```

File: robot-server/robot_server/service/routers/calibration_check.py (hunk 3 above)
Comment (sfoster1) [reply robot-server/robot_server/service/routers/calibration_check.py:39]: I think there's a larger question here - characterizing every possible action as an atomic `Command` either pushes a lot of work onto the client and isn't a good fit for all the different kinds of session, or becomes such a generic term it loses some of its usefulness. Take calibration check for instance - is "user thinks this is in the right place, is it in our error bounds?" an atomic command that wants the same kind of model as things like jog?

Comment (amitlissack) [reply robot-server/robot_server/service/routers/calibration_check.py:39]: What do you suggest?

**Comment by amitlissack** [comment]
> > I can't see the id being at the root level if we support collections. Unless you're thinking we return an object or a collection (?)
> 
> Yeah an `id` it wouldn't make sense in that context. If we wanted to flatten a little and be consistent with collections, I _guess_ I'm proposing something along the lines of:
> 
> ```js
> {
>   // ...
>   type: 'collection',
>   collection: [
>     {
>       type: 'item',
>       id: 'abc',
>       // ...
>     },
>     // ...
>   ]
> }
> ```

json api spec has data being a list or object. How do you feel about that? I like explicitness.

**Review by b-cooper — APPROVED** [review]
~This looks good to me. If we merge this first then I can update the few trigger-level changes in my pr (#5583) after rebasing and everything should work again.~

wrong PR moving this review/comment to #5581

**Comment by amitlissack** [comment]
This is being closed. These endpoints are now in edge and used to drive calibration check.

**Pull request closed without merging**